Some text field improvements.

deniz1a opened the pull request:
This is split from #8118.

Phrohdoh commented:
What do we gain from this? What is the reasoning behind it?

chrisforbes commented:
This seems vaguely OK to me, but I'd rather you pull the shared handler out into its own (possibly anonymous) function rather than invoking OnClick of another widget.

ScottNZ commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -109,8 +119,20 @@ public AssetBrowserLogic(Widget widget, Action onExit, World world)
+					closeButton.OnClick();
+				else
+				{
+					filenameInput.Text = null;
```
This seems dangerous to me. In general you want to use "" over null when it comes to UI text, and it looks to me like this could break if this handler runs twice.

penev92 replied:
Yup, always go for `string.Empty`.

deniz1a commented:
I put the shared closing statements in an Action variable and used that.

```filenameInput.Text = null;``` works because the set method of Text assigns empty string if the value is null.
https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Widgets/TextFieldWidget.cs#L23

But assigning an empty string directly is better I guess.

deniz1a commented:
This PR does the following:
* Asset browser: filter text field gets focus by default. Pressing Esc doesn't yield focus. Instead, if text field is empty, it closes the window. If text is not empty, it resets text (makes it empty) and updates filter.
* Lobby: Pressing Esc resets chat text field if it's not empty, otherwise it closes lobby window. In player name field, when Esc is pressed, text returns to unedited value and focus is yielded.
* Settings menu Display tab: For both frame limit and player name fields: pressing Esc resets text field and yields focus.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -108,9 +108,28 @@ public AssetBrowserLogic(Widget widget, Action onExit, World world)
+						player.Stop();
+					Ui.CloseWindow();
+					onExit();
+				};
```
Style nits: indented one level too far, and newline after the `};` please.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -218,13 +237,7 @@ public AssetBrowserLogic(Widget widget, Action onExit, World world)
-					Ui.CloseWindow();
-					onExit();
-				};
+				closeButton.OnClick = () => closeAssetBrowser();
```
You don't need the wrapper: `closeButton.OnClick = closeAssetBrowser;` does what you want.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -515,17 +515,16 @@ void CloseWindow()
+			Action closeLobby = () => { CloseWindow(); onExit(); };
 			var disconnectButton = lobby.Get<ButtonWidget>("DISCONNECT_BUTTON");
-			disconnectButton.OnClick = () => { CloseWindow(); onExit(); };
+			disconnectButton.OnClick = () => closeLobby();
```
`disconnectButton.OnClick = closeLobby`.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -549,6 +548,15 @@ void CloseWindow()
 				else
 					return true;
 			};
+			chatTextField.OnEscKey = () =>
```
Style nit: newline between `};` and this line.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -257,23 +257,28 @@ public static void SetupEditableNameWidget(Widget parent, Session.Slot s, Sessio
 
 			name.Text = c.Name;
-			name.OnLoseFocus = () =>
+			name.OnEnterKey = () =>
```
Removing the `OnLoseFocus` leaves the widget in an inconsistent state if you unfocus by clicking on the chat text field.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -257,23 +257,28 @@ public static void SetupEditableNameWidget(Widget parent, Session.Slot s, Sessio
+				name.YieldKeyboardFocus();
+				return true;
+			};
+			name.OnEscKey = () =>
```
Newline here too

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -190,14 +190,21 @@ Action InitDisplayPanel(Widget panel)
 				frameLimitTextfield.Text = ds.MaxFramerate.ToString();
+				frameLimitTextfield.YieldKeyboardFocus();
+				return true;
+			};
```
And after every other pseudo-block in this pr.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -190,14 +190,21 @@ Action InitDisplayPanel(Widget panel)
 			var frameLimitTextfield = panel.Get<TextFieldWidget>("FRAME_LIMIT_TEXTFIELD");
 			frameLimitTextfield.Text = ds.MaxFramerate.ToString();
-			frameLimitTextfield.OnLoseFocus = () =>
+			frameLimitTextfield.OnEnterKey = () =>
```
This has the same issue as the lobby.  If you unfocus using the mouse the widget text will no longer match the underlying data.

pchote commented:
There are some nice usability improvements here, but there are a few style issues and a couple of bugs that will need to be fixed before we can merge it.  Please also split this into individual commits for lobby / settings / asset browser with commit messages that describe what you said above.

deniz1a committed:
Fixes spelling in comment.

deniz1a committed:
Improves filter text field in asset browser.
Text field takes keyboard focus by default.
Esc key resets filter or closes window if text field is empty.

deniz1a committed:
Improves frame limit and player name text fields in settings menu.
Esc key resets text field to unedited value and yields focus.

deniz1a commented:
I fixed all the issues.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -550,6 +550,16 @@ void CloseWindow()
+			chatTextField.OnEscKey = () =>
+			{
+				if (chatTextField.Text.Length == 0)
+					closeLobby();
```
We deliberately chose not to do this because leaving the lobby is a high-impact action for something that can be so easily triggered.

deniz1a committed:
Improves player name text field in game lobby.
Esc key resets text field to unmodified value and yields focus.

deniz1a committed:
Resets chat text field in lobby when Esc is pressed.

obrakmann commented:
Looks fine now. :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/f3f5bdd6f80d3edf77bfdaa967e680cf969bc837%5E...f3f5bdd6f80d3edf77bfdaa967e680cf969bc837)
